scripts/lua/policies/security/oauth2.lua

```diff
@@ -55,37 +58,50 @@ function processWithRedis(red, securityObj)
   local token = {}
   local key = utils.concatStrings({"oauth:providers:", securityObj.provider, ":tokens:", accessToken})
   -- If we haven't cached the token go to the oauth provider
-  if not (redis.exists(red, key) == 1) then
-    token = exchange(accessToken, securityObj.provider)
-  else 
-    token = cjson.decode(redis.get(red, key))
+  if redis.exists(red, key) == 1 then
+    return cjson.decode(redis.get(red, key))
+  end
+ 
+  if securityObj.provider == 'facebook' or securityObj.provider == 'fake_facebook' then
```

Comment:
How can we keep custom provider logic out of the main oauth flow?

Should we keep all of the required info for a given provider in a single key/value pair? (e.g. for facebook, store both the accessToken and appToken together).

For Facebook, you now need to make two redis requests to validate the OAuth flow in addition to the stuff for routing.